text/0000-grid-primitives.md

```diff
@@ -0,0 +1,672 @@
+
+Standard integers are represented with the exponent set to zero. 
+
+### Enums
+
+An enum data type restricts values to a limited set of possible values. The
+definition for this data type includes a list of strings names describing a
+possible state of the enum. 
+
+### Structs
+
+A struct is a recursively defined collection of other named properties that
+represents two or more intrinsically linked values, like X/Y coordinates or RGB
+colors. These values can be of any Supply Chain data type, including STRUCT,
+allowing nesting to an arbitrary depth. Although versatile and powerful, structs
+are heavyweight and should be used conservatively; restrict struct use to
+linking values that must always be updated together. The transaction processor
+will enforce this usage, rejecting any transactions that do not have a value for
+every property in a struct.
+
+Note that although structs are built using a list of PropertyDefinitions, any
+nested use of the required property is meaningless and will be rejected by the
+transaction processor. As Properties are set in their entirety, either all of
+the struct is required or none of it is. In other words, partial structs are not
+allowed.
+
+### Locations
+
+Locations are represented as a pre-defined struct made up of a latitude,
+longitude pair.  Both latitude and longitude are represented as signed integers
+indicated millionths of degrees.
+
+## Schemas
+
+Property definitions are collected into a Schema data type, which defines all
+the possible properties for an item that belongs to a given schema.  These
+schemas include the following:
+
+- a name
+- a description
+- an owner
+- a list of `PropertyDefinitions`
+
+Schemas are used to validate a collection of property values defined in a data
+structure.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A schema and its property definitions are stored in state in the protobuf
+message format.  These messages are defined as follows:
+
+```
+message PropertyDefinition {
+    enum DataType {
+        UNSET_DATA_TYPE = 0;
+        BYTES = 1;
+        BOOLEAN = 2;
+        NUMBER = 3;
+        STRING = 4;
+        ENUM = 5;
+        STRUCT = 6;
+        LOCATION = 7;
+    }
+
+    // The name of the property
+    string name = 1;
+    // The data type of the value; must not be set to UNSET_DATA_TYPE.
+    DataType data_type = 2;
+    // Indicates that this is a required property in the Schema
+    bool required = 3;
+    // An optional description of the field.
+    string description = 4;
+
+    // The exponent for a NUMBER property
+    sint32 number_exponent = 10;
+    // The list of values for an ENUM property; must not be empty/ for
+    // properties of that type.
+    repeated string enum_options = 11;
+    // The list of property definitions for a STRUCT property; must  not be
+    // empty for properties of that type.
+    repeated PropertyDefinition struct_properties = 12;
+}
+
+message Schema {
```

Comment:
Do we need a SchemaList to handle (the very unlikely) hash collision?

Reply:
Added.